iib/workers/tasks/build_create_empty_index.py

```diff
@@ -0,0 +1,152 @@
+# SPDX-License-Identifier: GPL-3.0-or-later
+import logging
+import stat
+import tempfile
+import json
+import re
+
+from iib.workers.api_utils import set_request_state
+from iib.workers.tasks.build import (
+    _add_label_to_index,
+    _build_image,
+    _cleanup,
+    _create_and_push_manifest_list,
+    _get_index_database,
+    _opm_index_rm,
+    _push_image,
+    _serve_index_registry,
+    _update_index_image_build_state,
+    _update_index_image_pull_spec,
+)
+from iib.workers.tasks.celery import app
+from iib.workers.tasks.utils import (
+    chmod_recursively,
+    request_logger,
+    prepare_request_for_build,
+    run_cmd,
+    RequestConfigCreateIndexImage,
+)
+
+__all__ = ['handle_create_empty_index_request']
+
+log = logging.getLogger(__name__)
+
+
+def _get_present_operators(from_index, base_dir):
+    """
+    Get a list of operators already present in the index image.
+
+    :param str from_index: index image to inspect.
+    :param str base_dir: base directory to create temporary files in.
+    :return: list of unique present operators as provided by the grpc query
+    :rtype: list, list
+    :raises IIBError: if any of the commands fail.
+    """
+    db_path = _get_index_database(from_index, base_dir)
+    port, rpc_proc = _serve_index_registry(db_path)
+
+    # doc: operators https://www.redhat.com/ja/blog/deploying-operators-olm-bundles
+
+    operators = run_cmd(
+        ['grpcurl', '-plaintext', f'localhost:{port}', 'api.Registry/ListPackages'],
+        exc_msg='Failed to get operator data from index image',
+    )
+    rpc_proc.kill()
+
+    # If no data is returned there are not operators present
+    if not operators:
+        return []
+
+    # Transform returned data to parsable json
+    present_operators = []
+    new_operators = [json.loads(operator) for operator in re.split(r'(?<=})\n(?={)', operators)]
+
+    for operator in new_operators:
+        present_operators.append(operator['name'])
+
+    return present_operators
+
+
+@app.task
+@request_logger
+def handle_create_empty_index_request(
+    from_index, request_id, binary_image=None, labels=None, binary_image_config=None
+):
+    """Coordinate the the work needed to create the index image with labels.
+
+    :param str from_index: the pull specification of the container image containing the index that
+        the index image build will be based from.
+    :param int request_id: the ID of the IIB build request
+    :param str binary_image: the pull specification of the container image where the opm binary
+        gets copied from.
+    :param dict labels: the dict of labels required to be added to a new index image
+    :param dict binary_image_config: the dict of config required to identify the appropriate
+        ``binary_image`` to use.
+    """
+    _cleanup()
+
+    dist_scope = labels.get('com.redhat.index.delivery.distribution_scope', None)
```

Comment:
Can we default this to `prod`?

Reply:
sure

Reply:
This line is removed because if the distribution scope is specified in from_index it will fail. So I decided to not specify a distribution scope in _prepare_request_for_build and it will be modified at the newly build image later on.

Reply:
@yashvardhannanavati PTAL